### ISSUE TITLE
Consider using one of the error reporting libraries

### ISSUE THREAD

**Issue by 39555** [reubeno/brush#260]
One of the project's goals might be user-friendliness. We already have reedline, which is great. User friendly error would be the huge quality improvement over Bash.

- [ariadne](https://crates.io/crates/ariadne)
- [annotate-snippets](https://crates.io/crates/annotate-snippets)
- [codespan-reporting](https://crates.io/crates/codespan-reporting) looks unmaintained
- [miette](https://crates.io/crates/miette) to heavy, with a lot of features and proc macros. Used in Nushell
- [codesnake](https://crates.io/crates/codesnake) a simplified rewrite of `ariadne`
- [eyre](https://crates.io/crates/color-eyre) fork of anyhow with color reporting. I'm not sure about that.

Personally I like `annotate-snippets` the most for its simplicity and usage in `cargo`. Than `ariadne` looks good to me.

**Comment by reubeno** [comment]
Agreed on the opportunity for differentiated experience. I had previously looked into miette and was impressed with it. Would be great to get an analysis of the pros and cons of each of the candidate options you list. 

I'm also expecting that one of the challenges will be ensuring we have the right context for where the error is. (For example, our tokens have source info but it's hard to map our parser AST nodes back to where *they* came from if an error occurs while evaluating them.)

**Comment by reubeno** [comment]
@lu-zero I'm assigning this to you for now, since you've taken the initiative to dive in and we've converged on using `miette`.

Thank you! :smile:

**Comment by StudioLE** [comment]
miette is definitely the right choice here.

I've seen you already track the line number and character position therefore you should be able to take advantage of miette's [rich source annotations](https://github.com/zkat/miette?tab=readme-ov-file#example):

![](https://raw.githubusercontent.com/zkat/miette/main/images/serde_json.png)

![](https://raw.githubusercontent.com/zkat/miette/main/images/single-line-example.png)

**Comment by reubeno** [comment]
> miette is definitely the right choice here.
> 
> I've seen you already track the line number and character position therefore you should be able to take advantage of miette's [rich source annotations](https://github.com/zkat/miette?tab=readme-ov-file#example):
> 
> ![](https://raw.githubusercontent.com/zkat/miette/main/images/serde_json.png)
> 
> ![](https://raw.githubusercontent.com/zkat/miette/main/images/single-line-example.png)

Absolutely. I've prototyped using miette a few times but it was a bit too early and led to some earlier error type refactoring.

I think you're right that we have much of what we need -- just may need to require one of the diagnostic miette traits to be implemented by builtin errors.

It's definitely ready for someone to dive in and start putting something together 😄
